tests/MongoDB.Driver.Core.Tests/Jira/CSharp3173Tests.cs

```diff
@@ -0,0 +1,358 @@
+﻿/* Copyright 2020-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Linq;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+using FluentAssertions;
+using MongoDB.Bson;
+using MongoDB.Bson.TestHelpers.XunitExtensions;
+using MongoDB.Driver.Core.Clusters;
+using MongoDB.Driver.Core.Clusters.ServerSelectors;
+using MongoDB.Driver.Core.Configuration;
+using MongoDB.Driver.Core.ConnectionPools;
+using MongoDB.Driver.Core.Connections;
+using MongoDB.Driver.Core.Events;
+using MongoDB.Driver.Core.Helpers;
+using MongoDB.Driver.Core.Servers;
+using MongoDB.Driver.Core.WireProtocol.Messages;
+using MongoDB.Driver.Core.WireProtocol.Messages.Encoders;
+using Moq;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace MongoDB.Driver.Core.Tests.Jira
+{
+    public class CSharp3173Tests
+    {
+        private readonly static ClusterConnectionMode __clusterConnectionMode = ClusterConnectionMode.Sharded;
+        private readonly static ClusterId __clusterId = new ClusterId();
+        private readonly static EndPoint __endPoint1 = new DnsEndPoint("localhost", 27017);
+        private readonly static EndPoint __endPoint2 = new DnsEndPoint("localhost", 27018);
+        private readonly static TimeSpan __heartbeatInterval = TimeSpan.FromMilliseconds(200);
+        private readonly static ServerId __serverId1 = new ServerId(__clusterId, __endPoint1);
+        private readonly static ServerId __serverId2 = new ServerId(__clusterId, __endPoint2);
+
+        [Theory]
+        [ParameterAttributeData]
+        public void Ensure_command_network_error_before_hadnshake_is_correctly_handled([Values(false, true)] bool async, [Values(false, true)] bool streamable)
+        {
+            var eventCapturer = new EventCapturer().Capture<ServerDescriptionChangedEvent>();
+
+            // ensure that isMaster check response is finished only after network error
+            var hasNetworkErrorBeenTriggered = new TaskCompletionSource<bool>();
+            // ensure that there are no unexpected events between test ending and cluster disposing
+            var hasClusterBeenDisposed = new TaskCompletionSource<bool>();
+
+            EndPoint initialSelectedEndpoint = null;
+            using (var cluster = CreateAndSetupCluster(hasNetworkErrorBeenTriggered, hasClusterBeenDisposed, eventCapturer, streamable))
+            {
+                cluster._clusterId(__clusterId);
```

Comment:
I don't know how you can get away with this. The cluster implementation assumes that the ClusterId never changes.

One symptom of how this isn't really a good idea... during the whole test the ClusterId in the ClusterDescription does NOT match the ClusterId that you forcibly overwrote on this line.

I'm not sure why these mismatched ClusterIds don't cause problems, but it concerns me that we are forcibly overwriting the ClusterId value when the class is not expecting that.

In fact, the `_clusterId` field is declared **readonly** which only further emphasizes that it should not be changed.

Reply:
I'm not sure that the whole question is about the scope of this particular test (from my point of view, the target of the test is clearly defined and pretty specific) :) 

>The cluster implementation assumes that the ClusterId never changes.

I don't think that it's significantly important for emulating the particular test scenario. And anyway the clusterId is never changed **after** the time when the cluster has been initialized.

>One symptom of how this isn't really a good idea... during the whole test the ClusterId in the ClusterDescription does NOT match the ClusterId that you forcibly overwrote on this line.

I'm not sure that it's important for the scope of the current test, clusterId in ClusterDescription is used only in events. However, it's fixed in the latest commit.

>I'm not sure why these mismatched ClusterIds don't cause problems

most of the clusterId usage is about sdam logging.

>In fact, the _clusterId field is declared readonly which only further emphasizes that it should not be changed.

I don't think that this restriction is about mocking in the test. In fact, all of our reflection usages is a hack to avoid some code design restrictions.

Reply:
I wouldn't call using reflection in tests a hack... the alternative is to declare things that you normally want to be private internal or worse yet public, and that would be more like a hack.

I still think overwriting the `_clusterId` is not a good technique... and overwriting the `_description` also to compensate doesn't make it any better.

I understand why you are doing this... but I don't think it's the only approach or the best one. But in the interest of expediency I guess I'm willing to go along with this. I could try to find a better approach but not sure I have the time.

Reply:
>I wouldn't call using reflection in tests a hack...

it's a hack, in meaning that we break the design restriction. But it's not a hack in meaning that these restrictions are not created for test code or at very least they are not significantly important for the test code.

>but I don't think it's the only approach or the best one. 

I'm sure it's possible to do it in another way too, but I don't think that this makes any sense for the scope/target of this test :)

Reply:
Replace 63-64 with:

    ForceClusterId(cluster, __clusterId);

And add new helper method in THIS class:

    private void ForceClusterId(MultiServerCluster cluster, ClusterId clusterId)
    {
        Reflector.SetFieldValue(cluster, "_clusterId", clusterId);
        Reflector.SetFieldValue(cluster, "_description", ClusterDescription.CreateInitial(__clusterId, __clusterConnectionMode));
    }

Reply:
done